### PR TITLE
feat: add ACX rebate preview fee

### PR DESCRIPTION
**Pull request by james-a-morris**
This PR adds the new bridge fee table. It loads USD values to display in a non-blocking way. Additionally, the call to the referral summary endpoint is included in the table if it successfully loads. However, if the summary endpoint fails, then the UI will gracefully hide the row.
image

### CODE REVIEW

**Comment by james-a-morris** [comment]
Note: I am holding off on the CTA below as it should be ready to go in the recent deposit status update PR.
image

File: src/views/Bridge/components/BridgeForm.tsx [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { BigNumber } from "ethers";
 import styled from "@emotion/styled";
 
 import ExternalCardWrapper from "components/CardWrapper";
-import { PrimaryButton } from "components/Button";
+import { UnstyledButton } from "components/Button";
```
[the remark this answers is not in the document]
Comment (dohaki) [reply src/views/Bridge/components/BridgeForm.tsx:5]: There seem to be some inconsistencies introduced through this change
image

-> Wrong disabled state style

image

-> Wrong font color

File: src/views/Bridge/components/BridgeForm.tsx [hunk 2]
```diff
@@ -241,12 +247,14 @@ const ChangeAddressLink = styled(Text)`
...
+const Button = styled(UnstyledButton)`
+  display: flex;
+  flex-direction: row;
+  justify-content: center;
+  align-items: center;
 
-const Button = styled(PrimaryButton)`
+  background: #6cf9d8;
+  border-radius: 32px;
+  height: 64px;
```
Comment (dohaki) [inline src/views/Bridge/components/BridgeForm.tsx:258]: I think we can stick to the PrimaryButton component

File: src/hooks/useCoingeckoPrice.ts [hunk 3]
```diff
@@ -1,16 +1,44 @@
...
 ) {
-  return useQuery(
+  const query = useQuery(
     ["price", l1Token, baseCurrency],
     async () => getApiEndpoint().coingecko(l1Token, baseCurrency),
     {
       enabled,
     }
   );
+  const convertTokenToBaseCurrency = useCallback(
```
Comment (dohaki) [inline src/hooks/useCoingeckoPrice.ts:21]: Hmm, not so sure if I would expect this function to be part of the query hook 🤔

File: src/views/Bridge/components/BridgeForm.tsx [hunk 4]
```diff
@@ -49,6 +50,8 @@ type BridgeFormProps = {
   buttonLabel: string;
   isBridgeDisabled: boolean;
   validationError?: AmountInputError;
+  convertTokensToBaseCurrency: ConvertTokensToBaseCurrencyType;
```
Comment (dohaki) [inline src/views/Bridge/components/BridgeForm.tsx:53]: Can't we just import this function directly where it is needed inside the respective child component? This way we don't need to prop drill this function through the component tree.

File: src/views/Bridge/components/EstimatedTable.tsx [hunk 5]
```diff
@@ -178,3 +335,66 @@ const WarningInfoIcon = styled(InfoIcon)`
...
+  //Layout
+  display: flex;
+  padding: 6px 12px;
+  align-items: center;
+  gap: 4px;
+
+  // Style
+  border-radius: 22px;
+  border: 1px solid var(--color-interface-aqua-15, rgba(108, 249, 216, 0.15));
+  background: var(--color-interface-aqua-5, rgba(108, 249, 216, 0.05));
```
Comment (dohaki) [inline src/views/Bridge/components/EstimatedTable.tsx:366]: We should use the updated `COLORS` constant here

File: src/views/Bridge/components/EstimatedTable.tsx [hunk 6]
```diff
@@ -178,3 +335,66 @@ const WarningInfoIcon = styled(InfoIcon)`
...
+
+const Divider = styled.div`
+  height: 1px;
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  gap: 10px;
+  align-self: stretch;
+  background: var(--color-neutrals-grey-600, #3e4047);
```
Comment (dohaki) [inline src/views/Bridge/components/EstimatedTable.tsx:383]: We should use the updated COLORS constant here

File: src/views/Bridge/components/EstimatedTable.tsx [hunk 7]
```diff
@@ -178,3 +335,66 @@ const WarningInfoIcon = styled(InfoIcon)`
...
+  width: 100%;
+`;
+
+const ArrowIconDown = styled(UnstyledArrowIcon)`
+  path {
+    stroke: #9daab2;
```
Comment (dohaki) [inline src/views/Bridge/components/EstimatedTable.tsx:389]: We should use the updated COLORS constant here

File: src/views/Bridge/hooks/useBridge.ts [hunk 8]
```diff
@@ -72,6 +94,39 @@ export function useBridge() {
...
+    // Resolve total fees in terms of ACX
+    const totalFeesACX = totalFeesUSD
+      .mul(fixedPointAdjustment)
+      .div(acxExchangeRate);
+    // We can treat this as a percentage with 18 decimals
+    const feePct = parseUnits(referralSummary.referralRate.toString(), 18);
+    // Find the total referral reward as a percentage of the total fees
+    const referralReward = totalFeesACX.mul(feePct).div(fixedPointAdjustment);
+    // Return the referral reward in terms of ACX
+    return referralReward;
```
Comment (dohaki) [inline src/views/Bridge/hooks/useBridge.ts:122]: NIT: Again I think these comments are redundant and can be removed

File: src/views/Bridge/hooks/useBridge.ts [hunk 9]
```diff
@@ -41,6 +50,19 @@ export function useBridge() {
...
+
+  const { convertTokenToBaseCurrency: convertACXToBaseCurrency } =
+    useCoingeckoPrice(getToken("ACX").mainnetAddress!, "usd");
+
+  const convertTokensToBaseCurrency: ConvertTokensToBaseCurrencyType = {
+    l1: convertL1ToBaseCurrency,
+    acx: convertACXToBaseCurrency,
+  };
+
+  const { summary: referralSummary } = useReferralSummary(account);
```
Comment (dohaki) [inline src/views/Bridge/hooks/useBridge.ts:64]: Wdyt of moving this to the `<EstimatedTable />` component? I think this pattern of having a "god" hook that contains all hooks of the component tree defeats the main purpose of hooks:
- We still need to prop drill values down the component tree
- If a single value of these hooks updates, the whole tree needs to re-render

File: src/views/Bridge/hooks/useBridge.ts [hunk 10]
```diff
@@ -72,6 +94,39 @@ export function useBridge() {
     estimatedTime,
   } = usedTransferQuote || {};
 
+  const depositReferralReward = useMemo(() => {
```
Comment (dohaki) [inline src/views/Bridge/hooks/useBridge.ts:97]: We need to add some logic here as to whether a user is using a referral link of another user or if others used the referral link of the user 🤔 Because this impacts the rebate percentage as well.

**Review by dohaki — COMMENTED** [review]
Left some comments and suggestions! Let me know if something is unclear

File: src/views/Bridge/components/BridgeForm.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (james-a-morris) [reply src/views/Bridge/components/BridgeForm.tsx]: This was an issue with the `git merge`. It's corrected. Thanks for calling it out: https://github.com/across-protocol/frontend-v2/pull/896/commits/d1ce83f70d25bd09c3e8c2adeace4d06052f9db5

File: src/views/Bridge/components/BridgeForm.tsx (hunk 2 above)
Comment (james-a-morris) [reply src/views/Bridge/components/BridgeForm.tsx]: This was an issue with the git merge. It's corrected. Thanks for calling it out: https://github.com/across-protocol/frontend-v2/commit/d1ce83f70d25bd09c3e8c2adeace4d06052f9db5

File: src/views/Bridge/components/EstimatedTable.tsx (hunk 7 above)
Comment (james-a-morris) [reply src/views/Bridge/components/EstimatedTable.tsx]: https://github.com/across-protocol/frontend-v2/pull/896/commits/bd097749715d18019da4c98357fda26acb0ec36f

File: src/views/Bridge/components/EstimatedTable.tsx (hunk 6 above)
Comment (james-a-morris) [reply src/views/Bridge/components/EstimatedTable.tsx]: https://github.com/across-protocol/frontend-v2/pull/896/commits/bd097749715d18019da4c98357fda26acb0ec36f

File: src/views/Bridge/components/EstimatedTable.tsx (hunk 5 above)
Comment (james-a-morris) [reply src/views/Bridge/components/EstimatedTable.tsx]: https://github.com/across-protocol/frontend-v2/pull/896/commits/bd097749715d18019da4c98357fda26acb0ec36f

File: src/hooks/useCoingeckoPrice.ts [hunk 11]
```diff
@@ -1,16 +1,44 @@
...
+      // If the price is not defined yet, return undefined
+      if (!utils.isDefined(price) || !utils.isDefined(amount)) {
+        return undefined;
+      }
+      // Resolve token decimals
+      const { decimals } = getTokenByAddress(l1Token);
+      // Convert the amount to the base currency
+      const convertedAmount = ConvertDecimals(decimals, 18)(amount);
+      // Return the product of the price and the amount
+      return price.mul(convertedAmount).div(fixedPointAdjustment);
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply src/hooks/useCoingeckoPrice.ts]: Removed

File: src/hooks/useCoingeckoPrice.ts (hunk 3 above)
Comment (james-a-morris) [reply src/hooks/useCoingeckoPrice.ts]: Changed in favor of a new hook called `useTokenConversion`

File: src/views/Bridge/components/BridgeForm.tsx (hunk 4 above)
Comment (james-a-morris) [reply src/views/Bridge/components/BridgeForm.tsx]: Yep! Done

File: src/views/Bridge/hooks/useBridge.ts (hunk 8 above)
Comment (james-a-morris) [reply src/views/Bridge/hooks/useBridge.ts]: Removed

File: src/views/Bridge/hooks/useBridge.ts (hunk 9 above)
Comment (james-a-morris) [reply src/views/Bridge/hooks/useBridge.ts]: Yep! Absolutely

File: src/views/Bridge/hooks/useBridge.ts (hunk 10 above)
Comment (james-a-morris) [reply src/views/Bridge/hooks/useBridge.ts]: We now take into account the referral link provided. Per the docs, if a ref link is provided, the user gets 75% of the acx rewards which is 40% of the gas fee in ACX.

Comment (james-a-morris) [reply src/views/Bridge/hooks/useBridge.ts]: https://github.com/across-protocol/frontend-v2/pull/896/commits/9e2fd33e86835763a0176a0ebf6a71d2eca2b173

File: src/views/Bridge/components/EstimatedTable.tsx [hunk 12]
```diff
@@ -34,55 +73,166 @@ const EstimatedTable = ({
...
+    setIsDetailedFeesAvailable,
+    referralRewardAsBaseCurrency,
+    gasFeeAsBaseCurrency,
+    bridgeFeeAsBaseCurrency,
+    netFeeAsBaseCurrency,
+    formatUsd,
+    depositReferralReward,
+    depositReferralPercentage,
+    hasDepositReferralReward,
+  } = useEstimatedTable(token, gasFee, bridgeFee);
```
Comment (dohaki) [inline src/views/Bridge/components/EstimatedTable.tsx:88]: Nice, way cleaner than putting everything in the `useBridge` hook!

File: src/views/Bridge/hooks/useEstimatedTable.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    const totalFeesUSD = convertL1ToBaseCurrency(bridgeFee);
+    const acxExchangeRate = convertRewardToBaseCurrency(parseUnits("1", 18));
+    if (!utils.isDefined(totalFeesUSD) || !utils.isDefined(acxExchangeRate)) {
+      return undefined;
+    }
+    const totalFeesACX = totalFeesUSD
+      .mul(fixedPointAdjustment)
+      .div(acxExchangeRate);
+    // 75% of the reward goes to the referree of the referral rate
+    const availableRewardPercentage = referralSummary.referralRate * 0.75;
```
Comment (dohaki) [inline src/views/Bridge/hooks/useEstimatedTable.ts:46]: I think we also need to cover the case where the user self-refers. Then 100% of the referral rate goes to the user

**Review by dohaki — COMMENTED** [review]
Nice great work! Love it. I think we only need to cover one more case and we should be good to go

File: src/views/Bridge/hooks/useEstimatedTable.ts (hunk 13 above)
Comment (james-a-morris) [reply src/views/Bridge/hooks/useEstimatedTable.ts]: https://github.com/across-protocol/frontend-v2/pull/896/commits/eff02f9768f0eeb96d502179421bb8401ca077f7

**Comment by james-a-morris** [comment]
@dohaki rerequesting you - the changes that have been made are:
1. The staking CTA disappears when the user is not connected.
2. Reward fees are computed as a percentage of the entire fee.

File: src/views/Bridge/components/StakingCTA.tsx [hunk 14]
```diff
@@ -0,0 +1,124 @@
+import styled from "@emotion/styled";
+import { COLORS, QUERIESV2 } from "utils";
+import CloudBackground from "assets/bg-banners/cloud-staking.svg";
+import { ReactComponent as ACXLogo } from "assets/across.svg";
+import { PrimaryButton, Text } from "components";
+import copy from "copy-to-clipboard";
+import { useReferralLink } from "hooks/useReferralLink";
+import { useCallback } from "react";
+
+const StakingCTA = () => {
```
Comment (dohaki) [inline src/views/Bridge/components/StakingCTA.tsx:10]: NIT: The content of this CTA card is not really related to staking. Maybe changing the name to `ReferralCTA` would be better?

**Review by dohaki — APPROVED** [review]
Left one small nit on naming. But looks awesome otherwise!

File: src/hooks/useReferralLink.ts [hunk 15]
```diff
@@ -1,28 +1,21 @@
...
-
-  useEffect(() => {
-    const baseUrl = "across.to/?ref=";
-    const fullUrl = `${baseUrl}${account ?? ""}`;
-    const condensedUrl = `${baseUrl}${shortenString(account ?? "", "..", 4)}`;
-
-    setReferralLink(fullUrl);
-    setCondensedReferralLink(condensedUrl);
-    setReferralLinkWithProtocol(`https://${fullUrl}`);
+  return useMemo(() => {
```
Comment (dohaki) [inline src/hooks/useReferralLink.ts:7]: NIT: Do we need to use `useMemo` here? The computation doesn't seem that intensive to memoize the values 🤔

File: src/hooks/useReferrer.ts [hunk 16]
```diff
@@ -15,32 +16,32 @@ export default function useReferrer() {
   const [error, setError] = useState<string>("");
   useEffect(() => {
     setError("");
-    if (provider && r) {
+    if (isDefined(r)) {
       if (ethers.utils.isAddress(r)) {
         setIsResolved(true);
         return setAddress(r);
+      } else if (isDefined(provider)) {
```
Comment (dohaki) [inline src/hooks/useReferrer.ts:23]: NIT: If you change this to an `else if` then it would make sense to change line 22 to
```suggestion
        setAddress(r);
      } else if (isDefined(provider)) {
```

File: src/hooks/useRewardToken.ts [hunk 17]
```diff
@@ -0,0 +1,51 @@
+import { useConnection } from "hooks";
+import { useMemo } from "react";
+import { useQuery } from "react-query";
+import { getToken, parseUnits, rewardProgramsAvailable } from "utils";
+import useReferrer from "./useReferrer";
+import { useSimplifiedReferralSummary } from "./useSimplifiedReferralSummary";
+
+export function useRewardToken(destinationChainId: number) {
+  const rewardToken = useMemo(() => {
```
Comment (dohaki) [inline src/hooks/useRewardToken.ts:9]: NIT: Here again the question if we really need memoize the value 🤔

File: src/utils/constants.ts [hunk 18]
```diff
@@ -332,6 +333,31 @@ export const tokenList = [
...
+    rewardTokenSymbol: string;
+  }
+> = {
+  referrals: {
+    programName: "Across Referral Program",
+    primaryColor: "aqua",
+    url: "/rewards/referrals",
+    rewardTokenSymbol: "ACX",
+  },
+  "op-rebate": {
```
Comment (dohaki) [inline src/utils/constants.ts:353]: NIT: On the BE, we use the key `op-rebates`. So keeping it consistent wouldn't hurt

**Review by dohaki — APPROVED** [review]
Looking good 👍 Left only NITs and a few smaller suggestions. Feel free to tackle them in a separate PR.

File: src/utils/constants.ts (hunk 18 above)
Comment (james-a-morris) [reply src/utils/constants.ts]: Done! https://github.com/across-protocol/frontend-v2/pull/896/commits/6b965b209a8998fb22dd389369b5f66852840532

File: src/hooks/useReferrer.ts (hunk 16 above)
Comment (james-a-morris) [reply src/hooks/useReferrer.ts]: Yeah, that makes sense! This was old code that I didn't want to update, but the return statement needs to make more sense. https://github.com/across-protocol/frontend-v2/pull/896/commits/e377de57167f67edc32541d3f1f1700de59688ba

File: src/hooks/useRewardToken.ts (hunk 17 above)
Comment (james-a-morris) [reply src/hooks/useRewardToken.ts]: No that's a fair point. I wanted to memoize this because there's an `include`. On second look, this is going to be a small array so we can do away with it. https://github.com/across-protocol/frontend-v2/pull/896/commits/e5acc91a622dc8970e95ee363ac1ec714f18a410
